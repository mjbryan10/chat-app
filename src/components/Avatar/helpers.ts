export function stringToCaptialLetter(text: string): string {
  return text.charAt(0).toUpperCase();
}