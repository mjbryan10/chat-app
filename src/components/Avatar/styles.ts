import styled, {ThemeProps, Theme} from 'styled-components';

interface Props {
  theme: Theme;
  size?: number;
}

export const Container = styled.div`
  border: ${({theme}: ThemeProps<Theme>) => theme.border ?? 'inherit'};
  width: ${({size}:Props) => size}px;
  height: ${({size}:Props) => size}px;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const InnerContainer = styled.span`
  display: block;
  text-transform: uppercase;
`;
