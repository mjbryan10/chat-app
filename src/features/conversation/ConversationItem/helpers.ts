import { Conversation } from "shared/Api/@types";
import UserApi from "shared/Api/UserApi";

/**
 * Returns the conversation name if it is recorded.
 * 
 * If the name is not recorded, searches the coversation for another user,
 * and returns that users name if that is available.
 * 
 * If unable to find users name from API returns 'Anonymous'
 * 
 * Default returns 'Unknown conversation name' if all else fails.
 * @param conversation The conversation the name relates to
 * @param currentUserId The current user requesting the conversation name
 */
export const computeConversationName = async (
  conversation: Conversation,
  currentUserId: number | null
): Promise<string> => {
  const { name } = conversation.conversation;

  if (name) return name;

  const friendUser = conversation.users.find((user) => user.userid !== currentUserId);

  if (friendUser) {
    try {
      return await new UserApi().fetchUserNameById(friendUser.userid)
    } catch (error) {
      return 'Anonymous'
    } //TODO: Test if this works!
    
    //  const response = await new UserApi().fetchUserNameById(friendUser.userid)
     

    //  return response.data ? response.data.name : 'Anonymous';
  }
  return 'Unknown conversation name';
};
/**
 * Compares a conversation to the one in the Redux store and returns `true` if this conversation
 * is the selected on in the redux store.
 * 
 * Returns `false` if otherwise.
 * @param conversation the conversation to compare to selected conversation
 * @param currentConversationId The ID of the current conversation selected in Redux store
 */
export const isSelectedConversation = (
  conversation: Conversation,
  currentConversationId: number | null
): boolean => {
  if (!currentConversationId) return false;
  if (conversation.conversation.conversationId === currentConversationId) return true;
  return false;
};