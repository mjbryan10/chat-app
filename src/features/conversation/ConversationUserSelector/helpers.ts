import { User } from "shared/Api/types";

/**
 * Sorts an array of users alphabetically by name.
 * @param users An array of users to sort by name
 */
export const sortUsers = (users: User[]) => {
   if (!users.length) return [];
   return users.slice().sort((a, b) => a.name.localeCompare(b.name));
};
