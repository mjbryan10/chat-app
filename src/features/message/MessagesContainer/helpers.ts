import { Message } from "shared/Api/@types";
import { colorSpectrumArray } from "shared/theme/@types";
import UserApi from "shared/Api/UserApi";
import { Participant } from "../messageSlice";

/**
 * A function for gathering and generating information relating to each user
 * participant of a conversation.
 *
 * Fetches the participants name from the API.
 *
 * Generates a color for each participant from the Spectrum array.
 * Informs if the participant is the current user, to highlight if message owner.
 *
 * @param messages The messages from which to gather participant ids
 * @param participants An array of users participating in the given conversation.
 * @param currentUserId The currentUser viewing the conversation.
 * @returns Participant[] an array of objects containg participant details
 */
export const generateParticipants = (
   messages: Message[],
   currentUserId: number,
   participants: Participant[] = []
) => {
   // this should look at messages, if new user fetch name and assign a color.
   const colors = colorSpectrumArray;
   let colorsIndex = 0;
   if (participants.length) {
      //if participants already populated, set index to last color used + 1.
      const lastColorUsed = participants[participants.length - 1].color;
      colorsIndex = colors.findIndex((color) => color === lastColorUsed) + 1;
   }
   const userApi = new UserApi();
   messages.forEach((message) => {
      if (!participants.filter((user) => user.id === message.senderId).length) {
         let username = 'Anonymous';
         userApi
            .fetchUserNameById(message.senderId)
            .then((result) => {
               username = result; //Change username if appropiate.
            })
            .finally(() => {
               const newParticipant = {
                  id: message.senderId,
                  name: username,
                  color: colors[colorsIndex] || colors[0],
                  isOwner: message.senderId === currentUserId,
               };
               participants.push(newParticipant);

               colorsIndex >= colors.length ? (colorsIndex = 0) : colorsIndex++;
               //cycle to next available color (index)
            });
      }
   });
   return participants;
};