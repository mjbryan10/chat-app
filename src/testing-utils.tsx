import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }