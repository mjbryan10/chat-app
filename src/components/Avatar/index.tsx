import React, { FC, useMemo } from 'react';
import * as S from './styles';
import { ThemeColor } from 'styled-components';

export interface Props {
   title?: string;
   size?: number
   color?: ThemeColor;
}

export function stringToCaptialLetter(text: string): string {
  return text.charAt(0).toUpperCase();
}

const Avatar: FC<Props> = ({ title = 'anonymous', size = 50, color = "text" }) => {
  const letterMemo = useMemo(() => stringToCaptialLetter(title), [title])
   return (
      <S.Container size={size} color={color}>
         <S.InnerContainer size={size} data-testid="avatar-inner-container">{letterMemo}</S.InnerContainer>
      </S.Container>
   );
};

export default Avatar;
