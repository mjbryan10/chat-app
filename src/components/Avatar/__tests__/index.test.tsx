import React from 'react';
import Avatar from '..';
import { render } from 'testing-utils';
import { stringToCaptialLetter } from '../helpers';

describe('Avatar component', () => {
   describe('stringToCaptialLetter', () => {
      test('should return the first letter captalised', () => {
         const testText = 'test';

         const result = stringToCaptialLetter(testText);

         expect(result).toEqual('T');
      });
   });

   test('should render an Avatar with `A` if no text provided', () => {
      const utils = render(<Avatar />);

      const innerContainer = utils.getByTestId('avatar-inner-container');

      expect(innerContainer.innerHTML).toEqual('A');
   });

   test('should render the first letter captalised only', () => {
      const utils = render(<Avatar title="Demo name or message" />);

      const innerContainer = utils.getByTestId('avatar-inner-container');

      expect(innerContainer.innerHTML).toEqual('D');
   });

   test('should render correctly', () => {
      const { asFragment } = render(<Avatar />);

      expect(asFragment()).toMatchSnapshot();
   });
});
