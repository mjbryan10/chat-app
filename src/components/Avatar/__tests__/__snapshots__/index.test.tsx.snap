// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm kLuhel"
    color="text"
    size="50"
  >
    <span
      class="sc-AxirZ dsXsoK"
      data-testid="avatar-inner-container"
      size="50"
    >
      A
    </span>
  </div>
</DocumentFragment>
`;
